src/tracks/ui/TextEditHelper.cpp

```diff
@@ -0,0 +1,666 @@
+    case WXK_RETURN:
+    case WXK_NUMPAD_ENTER:
+    case WXK_TAB:
+        delegate->OnTextEditFinished(project, mText);
+        return true;
+    }
+    return false;
+}
+
+bool TextEditHelper::OnChar(int charCode, AudacityProject* project)
+{
+    auto delegate = mDelegate.lock();
+    if (!delegate)
+        return false;
+
+    if (charCode == 0 || wxIscntrl(charCode)) {
+        return false;
+    }
+
+    // Test if cursor is in the end of string or not
+    if (mInitialCursorPos != mCurrentCursorPos)
+        RemoveSelectedText(project);
+
+    if (mCurrentCursorPos < (int)mText.length()) {
+        // Get substring on the righthand side of cursor
+        wxString rightPart = mText.Mid(mCurrentCursorPos);
+        // Set title to substring on the lefthand side of cursor
+        mText = mText.Left(mCurrentCursorPos);
+        //append charcode
+        mText += charCode;
+        //append the right part substring
+        mText += rightPart;
+    }
+    else
+        //append charCode
+        mText += charCode;
+
+    delegate->OnTextModified(project, mText);
+
+    //moving cursor position forward
+    mInitialCursorPos = ++mCurrentCursorPos;
+
+    return true;
+}
+
+bool TextEditHelper::OnClick(const wxMouseEvent& event, AudacityProject*)
+{
+    if (event.ButtonDown())
+    {
+        if (mBBox.Contains(event.GetPosition()))
+        {
+            if (event.LeftDown()) 
+            {
+                mRightDragging = false;
+                auto position = FindCursorIndex(event.GetPosition());
+                auto initial = mInitialCursorPos;
+                if (event.ShiftDown()) {
+#ifdef __WXMAC__
+                    // Set the drag anchor at the end of the previous selection
+                    // that is farther from the NEW drag end
+                    const auto current = mCurrentCursorPos;
+                    if (abs(position - current) > abs(position - initial))
+                        initial = current;
+#else
+                    // initial position remains as before
+#endif
+                }
+                else
+                    initial = position;
+
+                mInitialCursorPos = initial;
+                mCurrentCursorPos = position;
+            }
+            else
+            {
+                if (mInitialCursorPos == mCurrentCursorPos)
+                {
+                    auto position = FindCursorIndex(event.GetPosition());
+                    mInitialCursorPos = mCurrentCursorPos = position;
+                }
+                // Actually this might be right or middle down
+                mRightDragging = true;
+            }
+
+#if defined(__WXGTK__) && (HAVE_GTK)
+            if (evt.MiddleDown()) {
+                // Paste text, making a NEW label if none is selected.
+                wxTheClipboard->UsePrimarySelection(true);
+                view.PasteSelectedText(project, newSel.t0(), newSel.t1());
+                wxTheClipboard->UsePrimarySelection(false);
```

Comment:
I don't know if it is important, but I think this if-block needs to move down, outside one more } , to correct the analogy with LabelTextHandle::HandleTextClick:  Do it whether bounding box contains the position, or not.

Reply:
Agreed, thanks!